### ISSUE TITLE
[intro.object] CWG2527 might be incorrectly closed as NAD

### ISSUE THREAD

**Issue by frederick-vs-ja** [cplusplus/CWG#254]
Full name of submitter (unless configured in github; will be published with the issue): Jiang An

Reference (section label): [intro.object]

Link to reflector thread (if any):

Issue description:
The original description of CWG2527 (which was too terse to address my intent) said:
> The definition of a “potentially-overlapping subobject” in 6.7.2 [[intro.object](https://eel.is/c++draft/intro.object)] paragraph 7 does not exclude non-class subobjects; in particular, 9.12.11 [[dcl.attr.nouniqueaddr](https://eel.is/c++draft/dcl.attr.nouniqueaddr)] makes no restrictions on the types of members declared with the `no_unique_address` attribute. It is not clear that a potentially-overlapping scalar member or array of scalar elements is useful. Should there be a restriction on the type of potentially-overlapping subjects?

And the response was:
> Restricting the type of a potentially-overlapping subobject would make it difficult to use `no_unique_address` on a subobject of dependent type, which may be a non-class type in some, but not all, specializations. Compilers can warn about non-sensical uses in non-dependent contexts.

My original opinions were:
1. Array element subobjects can't be potentially-overlapping subobjects, so layout compaction can't be enabled by making an array member subobject potentially-overlapping. However, this may bring some kinds of UB because of [[basic.life]/8.4](https://eel.is/c++draft/basic.life#8.4).
   1. The wording might be fixed by CWG2677, but the undefined cases won't be changed.
2. Currently no known implementation reuses padding bytes in a scalar subobject.

Furtherly, my intentional sugesstion was not restricting use of `no_unique_address`, but only making it effectless in more cases. This should not make use of `no_unique_address` difficult. Note that `no_unique_address` is already allowed to be used with non-static reference data member and it is simply ignored in such case, because a non-static reference member is not a subobject.

Suggested resolution:

Option A:
Change [[intro.object]/7.2](https://eel.is/c++draft/intro.object#7.2) as indicated:
> a non-static data member <ins>**that is of a non-array object type and**</ins> declared with the `no_unique_address` attribute.

Option B:
Change [[intro.object]/7.2](https://eel.is/c++draft/intro.object#7.2) as indicated:
> a non-static data member <ins>**that is of a class type and**</ins> declared with the `no_unique_address` attribute.

Drafting note: no restriction should be added to the use of `no_unique_address` attribute.

**Comment by jensmaurer** [comment]
"no_unique_address" means simply that the programmer gives up some guarantees he/she might otherwise get from the implementation.

If the programmer states that he/she gives up some guarantees on a subobject of scalar or array type, why should we not take this statement for granted?

For example, one could envision an implementation with char = 8 bits, int = 24 bits aligned to 32 bits, so for the scalar "int" or "array of int" cases, there are 8 bits of tail padding that may be reused. The fact that such an implementation is currently unknown is not rationale enough to tighten the rules of the standard, I think.

In short, the wording in the current Working Draft seems clear as written; I can't see a defect.
If you wish to pursue a change in the rules, please write a paper targeted at EWG, with rationale.

**Comment by frederick-vs-ja** [comment]
> For example, one could envision an implementation with char = 8 bits, int = 24 bits aligned to 32 bits, so for the scalar "int" or "array of int" cases, there are 8 bits of tail padding that may be reused.

Oh, if the intent is that `no_unique_address` allows reusing tail padding of arrays, then I believe the defect is that we don't say array element subobjects may be potentially-overlapping.

It seems that currently no padding of array can be reused because the padding must belong to some element subobject which is never potentially-overlapping.

**Comment by jensmaurer** [comment]
> It seems that currently no padding of array can be reused because the padding must belong to some element subobject which is never potentially-overlapping.

I think you can consider the tail padding on the final array element as being the tail padding of the entire array. I think that same reasoning applies to nested structs:

```
struct X {
  int x; // 4 bytes
  char y;  // 1 byte
  // tail padding of 3 bytes, I think
};

struct S {
   struct X x;
   // implied tail padding of 3 bytes, I think
};

struct Y {
  [[no_unique_address]] S s;
  [[no_unique_address]] char z;  // can be placed in tail padding of "s", I hope
};
```

In any case, a change to the words here appear to be feature requests and not defect reports. The language rules might not be as orthogonal as one might hope, but that doesn't imply there's a defect.

**Comment by frederick-vs-ja** [comment]
> I think you can consider the tail padding on the final array element as being the tail padding of the entire array. I think that same reasoning applies to nested structs:
> 
> ```
> struct X {
>   int x; // 4 bytes
>   char y;  // 1 byte
>   // tail padding of 3 bytes, I think
> };
> 
> struct S {
>    struct X x;
>    // implied tail padding of 3 bytes, I think
> };
> 
> struct Y {
>   [[no_unique_address]] S s;
>   [[no_unique_address]] char z;  // can be placed in tail padding of "s", I hope
> };
> ```
> 
> In any case, changing the words here appears to be a feature request and not a defect report. The language rules might not be as orthogonal as one might hope, but that doesn't imply there's a defect.

The potentially-overlapping property doesn't seem implicitly propagated to subobjects, which may defeat your intent and is possibly a defect.

In this example, given a `Y` complete object `y`, IIUC we only specify that `y.s` and `y.z` are potentially-overlapping subobjects, while `y.s.x` is not (because it is a member subobject where the member is not declared with `[[no_unique_address]]`), so the tail padding of `y.s.x` cannot be reused for `y.z`.
